### PR TITLE
Add "router fw" commands to minirouter to add iptables FORWARD rules

### LINKED ISSUE #1236: minirouter: any plans for simple firewall support?
#344 mentioned including a simple firewall in the container based router, but iptables isn't present in the defacto `uminirouter` container root filesystem and the `router` API doen't include any subcommands for firewall rules.

Just curious if there's any plans for this? I'm fine with creating a custom minirouter configuration that includes `iptables` and adding rules via `cc exec`. Was just curious if there were any plans for firewall support in the API in the future.

### PR DESCRIPTION
**Pull request by activeshadow**
Both minimega and minirouter were updated to support configuring
iptables in the minirouter instance using new "router fw" commands.

minirouter was updated to include support for "fw" commands for both
global iptables rules and grouped rules using iptables chains.

minimega was updated to include support for "router <vm> fw" commands
that generate the minirouter "fw" commands in the config script sent to
the minirouter instance via miniccc.

Documentation for minimega's "router" command was also updated to
include details and examples of how to use the new "fw" commands.

closes #1236

### CODE REVIEW

File: src/minimega/router.go [hunk 1]
```diff
@@ -945,6 +1026,84 @@ func (r *Router) RouteBGPDel(processname string, local, clearall bool) error {
...
 
+func (r *Router) FirewallDefault(d string) error {
+	log.Debug("RouterFirewallDefault: %s", d)
+
+	r.FW.defaultAction = d
+	return nil
+}
+
+func (r *Router) FirewallAdd(n int, in bool, src, dst, proto, action string) error {
+	log.Debug("RouterFirewallAdd: %d, %b, %s, %s, %s, %s", n, in, src, dst, proto, action)
```
Comment (aherna) [inline src/minimega/router.go:1037]: Not getting past vetting run compinling ,the second parameter %b
@activeshadow

File: src/minimega/router.go [hunk 2]
```diff
@@ -945,6 +1026,84 @@ func (r *Router) RouteBGPDel(processname string, local, clearall bool) error {
...
+	}
+
+	rule := &fwRule{src: src, dst: dst, proto: proto, action: action}
+	c.rules = append(c.rules, rule)
+
+	return nil
+}
+
+func (r *Router) FirewallChainApply(n int, in bool, chain string) error {
+	log.Debug("RouterFirewallChainApply: %d, %b, %s", n, in, chain)
```
Comment (aherna) [inline src/minimega/router.go:1078]: Not getting past vetting run compinling ,the second parameter %b
@activeshadow

File: src/minirouter/fw.go [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+			log.Errorln(err)
+		}
+
+		return
+	}
+
+	if c.BoolArgs["default"] {
+		if c.BoolArgs["accept"] {
+			if out, err := exec.Command("iptables", "-P", "FORWARD", "ACCEPT").CombinedOutput(); err != nil {
+				log.Errorln("defaulting FORWARD to ACCEPT %v: %v", err, string(out))
```
Comment (aherna) [inline src/minirouter/fw.go:177]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+				return
+			}
+		} else if c.BoolArgs["drop"] {
+			if out, err := exec.Command("iptables", "-P", "FORWARD", "DROP").CombinedOutput(); err != nil {
+				log.Errorln("defaulting FORWARD to DROP %v: %v", err, string(out))
```
Comment (aherna) [inline src/minirouter/fw.go:182]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+				return
+			}
+		} else if c.BoolArgs["reject"] {
+			if out, err := exec.Command("iptables", "-P", "FORWARD", "REJECT").CombinedOutput(); err != nil {
+				log.Errorln("defaulting FORWARD to REJECT %v: %v", err, string(out))
```
Comment (aherna) [inline src/minirouter/fw.go:187]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+		var (
+			idx   = -1
+			iface = "lo"
+			err   error
+		)
+
+		if c.StringArgs["index"] != "lo" {
+			idx, err = strconv.Atoi(c.StringArgs["index"])
+			if err != nil {
+				log.Errorln("converting interface index: %v", err)
```
Comment (aherna) [inline src/minirouter/fw.go:205]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 7]
```diff
@@ -0,0 +1,364 @@
...
+				return
+			}
+		}
+
+		if idx != -1 {
+			// get interface name using the index
+			if iface, err = findEth(idx); err != nil {
+				log.Errorln("getting interface name for index: %v", err)
```
Comment (aherna) [inline src/minirouter/fw.go:213]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 8]
```diff
@@ -0,0 +1,364 @@
...
+
+		if src := c.StringArgs["src"]; src != "" {
+			fields := strings.Split(src, ":")
+
+			switch len(fields) {
+			case 1:
+				rule = append(rule, "-s", src)
+			case 2:
+				if _, err = strconv.Atoi(fields[1]); err != nil {
+					log.Errorln("converting source port: %v", err)
```
Comment (aherna) [inline src/minirouter/fw.go:238]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

File: src/minirouter/fw.go [hunk 9]
```diff
@@ -0,0 +1,364 @@
...
+
+		if dst := c.StringArgs["dst"]; dst != "" {
+			fields := strings.Split(dst, ":")
+
+			switch len(fields) {
+			case 1:
+				rule = append(rule, "-d", dst)
+			case 2:
+				if _, err = strconv.Atoi(fields[1]); err != nil {
+					log.Errorln("converting destination port: %v", err)
```
Comment (aherna) [inline src/minirouter/fw.go:262]: Formatting directive error during vet. Use  log.Error(....) instead
Vet error "Errorln call has possible formatting directive %v"

**Review by aherna — CHANGES_REQUESTED** [review]
the unit test for minirouter fails, and go vetting fails. @activeshadow

File: src/minimega/router.go [hunk 10]
```diff
@@ -341,6 +368,60 @@ func (r *Router) writeConfig(w io.Writer) error {
...
+			if rule.src != "" {
+				cmd = fmt.Sprintf("%s %s", cmd, rule.src)
+			}
+
+			cmd = fmt.Sprintf("%s %s %s", cmd, rule.dst, rule.proto)
+
+			fmt.Fprintln(w, cmd)
+		}
+
+		fmt.Fprintf(w, "fw chain %s default action %s\n", name, chain.defaultAction)
```
Comment (aherna) [inline src/minimega/router.go:389]: Reject doesnt seem to work?
ERROR fw.go:187: defaulting FORWARD to REJECT exit status 1: iptables: Bad policy name. Run `dmesg' for more information.
@activeshadow

**Review by aherna — CHANGES_REQUESTED** [review]

File: src/minimega/router_cli.go [hunk 11]
```diff
@@ -127,6 +151,13 @@ router takes a number of subcommands:
 			"router <vm> <route,> <bgp,> <processname> <rrclient,>",
 			"router <vm> <route,> <bgp,> <processname> <export,> <all,filter> <filtername>",
 			//"router <vm> <importbird,> <configfilepath>", TODO
+			"router <vm> <fw,> <default,> <accept,drop,reject>",
```
[the remark this answers is not in the document]
Comment (activeshadow) [reply src/minimega/router_cli.go]: @aherna This has been fixed...

```
$> mm router rtr
...
Firewall Rules:
  (iface 0) outgoing: <any> --> 192.168.0.5:80 via tcp -> accept
  (iface 0) outgoing: <any> --> <any> via <any> -> allow-http
  Default Action: drop
Firewall Chains:
  allow-http: <any> --> 192.168.0.5:80 via tcp -> accept
  Default Action (allow-http): drop
...
```

File: src/minimega/router.go [hunk 12]
```diff
@@ -341,6 +368,60 @@ func (r *Router) writeConfig(w io.Writer) error {
 	fmt.Fprintf(w, "bird routerid %v\n", r.routerID)
 	fmt.Fprintf(w, "bird commit\n")
 
+	// ***** firewall stuff ***** //
+
+	fmt.Fprintln(w, "fw flush") // no need to manage firewall state - just start over
```
[the remark this answers is not in the document]
Comment (activeshadow) [reply src/minimega/router.go]: @aherna this has been fixed.

File: src/minimega/router.go (hunk 10 above)
Comment (activeshadow) [reply src/minimega/router.go]: @aherna this has been fixed... `REJECT` cannot be used as a default chain policy. To actively reject packets, the last rule in a chain needs to use `REJECT` as a target. As such, I've removed it as an option for the default firewall policy.

**Comment by activeshadow** [comment]
@aherna any luck getting to review my updates?

**Review by aherna — APPROVED** [review]
LGTM

**Pull request merged**